File: 2024/5.py
import functools
import itertools

import util


def part1(input: str) -> int:
    lines = input.splitlines()
    i = 0
    rules = set()
    while (line := lines[i]) != "":
        a, b = line.split("|")
        rules.add((a, b))
        i += 1

    s = 0
    i += 1
    while i < len(lines):
        update = lines[i].split(",")
        if all((a, b) in rules for a, b in itertools.combinations(update, 2)):
            s += int(update[len(update) // 2])
        i += 1

    return s


def part2(input: str) -> int:
    lines = input.splitlines()
    i = 0
    rules = set()
    while lines[i] != "":
        a, b = lines[i].split("|")
        rules.add((a, b))
        i += 1

    def cmp(a, b):
        return -1 if (a, b) in rules else 1

    s = 0
    i += 1
    while i < len(lines):
        update = lines[i].split(",")
        if any((a, b) not in rules for a, b in itertools.combinations(update, 2)):
            update = sorted(
                update,
                key=functools.cmp_to_key(cmp),
            )
            s += int(update[len(update) // 2])

        i += 1

    return s


if __name__ == "__main__":
    util.run(5, part1, part2)

File: 2025/5.ts
import { run } from "./util";

async function part1(input: string): Promise<number> {
  const parts = input.split("\n\n");
  const ranges = parts[0]
    .split("\n")
    .map((line) => line.split("-").map(Number));
  const ids = parts[1].split("\n").map(Number);
  let n = 0;
  for (const id of ids) {
    for (const range of ranges) {
      if (id >= range[0] && id <= range[1]) {
        n++;
        break;
      }
    }
  }
  return n;
}

async function part2(input: string): Promise<number> {
  const ranges = input
    .split("\n\n")[0]
    .split("\n")
    .map((line) => line.split("-").map(Number))
    .sort((a, b) => a[0] - b[0]);

  let n = 0;
  let curr_max = 0;

  for (let [l, r] of ranges) {
    if (r >= curr_max) {
      n += r - Math.max(l, curr_max) + 1;
      curr_max = r + 1;
    }
  }
  return n;
}

run(5, part1, part2);
